File: Regresion_Lineal/regresion_lineal.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <mysql.h>

// Valores por defecto (entorno local XAMPP). Se pueden sobreescribir con
// variables de entorno para no exponer credenciales en el codigo fuente.
#define DEF_HOST     "127.0.0.1"
#define DEF_USER     "root"
#define DEF_PASSWORD ""
#define DEF_DATABASE "mineria_datos"
#define DEF_PORT     3306
#define PLUGIN_DIR   "C:\\Program Files\\MySQL\\MySQL Workbench 8.0 CE"

// Devuelve el valor de la variable de entorno o el valor por defecto.
static const char *env_o_def(const char *nombre, const char *valor_def) {
    const char *v = getenv(nombre);
    return (v && *v) ? v : valor_def;
}

typedef struct {
    char mes[30];
    double inversion;
    double ventas;
} Dato;

void regresion_lineal(Dato *datos, int n, double *b0, double *b1) {
    double suma_x = 0, suma_y = 0, suma_xy = 0, suma_x2 = 0;

    for (int i = 0; i < n; i++) {
        double x = datos[i].inversion;
        double y = datos[i].ventas;
        suma_x  += x;
        suma_y  += y;
        suma_xy += x * y;
        suma_x2 += x * x;
    }

    *b1 = (n * suma_xy - suma_x * suma_y) / (n * suma_x2 - suma_x * suma_x);
    *b0 = (suma_y - (*b1) * suma_x) / n;
}

void metricas(Dato *datos, int n, double b0, double b1,
              double *r2, double *mse, double *rmse) {
    double suma_y = 0;
    for (int i = 0; i < n; i++) suma_y += datos[i].ventas;
    double y_media = suma_y / n;

    double ss_tot = 0, ss_res = 0;
    for (int i = 0; i < n; i++) {
        double y     = datos[i].ventas;
        double y_est = b0 + b1 * datos[i].inversion;
        ss_tot += (y - y_media) * (y - y_media);
        ss_res += (y - y_est) * (y - y_est);
    }

    *r2   = (ss_tot > 0) ? 1.0 - ss_res / ss_tot : 1.0;
    *mse  = ss_res / n;
    *rmse = sqrt(*mse);
}

int main() {
    MYSQL *conn = mysql_init(NULL);
    if (!conn) {
        fprintf(stderr, "Error al inicializar MySQL\n");
        return 1;
    }

    mysql_options(conn, MYSQL_PLUGIN_DIR, PLUGIN_DIR);

    const char *host = env_o_def("DB_HOST", DEF_HOST);
    const char *user = env_o_def("DB_USER", DEF_USER);
    const char *pass = env_o_def("DB_PASS", DEF_PASSWORD);
    const char *db   = env_o_def("DB_NAME", DEF_DATABASE);
    const char *port_s = getenv("DB_PORT");
    unsigned int port = (port_s && *port_s) ? (unsigned int)atoi(port_s) : DEF_PORT;

    if (!mysql_real_connect(conn, host, user, pass, db, port, NULL, 0)) {
        fprintf(stderr, "Error de conexion: %s\n", mysql_error(conn));
        mysql_close(conn);
        return 1;
    }

    if (mysql_query(conn, "SELECT mes, inversion, ventas FROM inversion_ventas ORDER BY id")) {
        fprintf(stderr, "Error en consulta: %s\n", mysql_error(conn));
        mysql_close(conn);
        return 1;
    }

    MYSQL_RES *resultado = mysql_store_result(conn);
    if (!resultado) {
        fprintf(stderr, "Error al obtener resultado: %s\n", mysql_error(conn));
        mysql_close(conn);
        return 1;
    }

    int n = (int)mysql_num_rows(resultado);
    Dato *datos = malloc(n * sizeof(Dato));

    printf("========================================\n");
    printf("  DATOS DE LA BASE DE DATOS\n");
    printf("========================================\n");
    printf("%-12s %14s %12s\n", "Mes", "Inversion (X)", "Ventas (Y)");
    printf("----------------------------------------\n");

    MYSQL_ROW fila;
    int i = 0;
    while ((fila = mysql_fetch_row(resultado))) {
        strncpy(datos[i].mes, fila[0], sizeof(datos[i].mes) - 1);
        datos[i].inversion = atof(fila[1]);
        datos[i].ventas    = atof(fila[2]);
        printf("%-12s %14.2f %12.2f\n", datos[i].mes, datos[i].inversion, datos[i].ventas);
        i++;
    }

    mysql_free_result(resultado);
    mysql_close(conn);

    double b0, b1;
    regresion_lineal(datos, n, &b0, &b1);

    printf("\n========================================\n");
    printf("  REGRESION LINEAL\n");
    printf("========================================\n");
    printf("Intercepto (b0): %.4f\n", b0);
    printf("Pendiente  (b1): %.4f\n", b1);
    printf("Ecuacion:  Y = %.2f + %.2f * X\n", b0, b1);

    double r2, mse, rmse;
    metricas(datos, n, b0, b1, &r2, &mse, &rmse);

    printf("\n========================================\n");
    printf("  METRICAS DE BONDAD DE AJUSTE\n");
    printf("========================================\n");
    printf("R^2  (determinacion): %.4f  (%.2f%% explicado)\n", r2, r2 * 100);
    printf("MSE  (error cuad. medio): %.4f\n", mse);
    printf("RMSE (raiz del MSE): %.4f\n", rmse);

    printf("\n========================================\n");
    printf("  PREDICCIONES\n");
    printf("========================================\n");
    double valores_pred[] = {10, 20, 30, 40, 50, 60};
    int num_pred = sizeof(valores_pred) / sizeof(valores_pred[0]);
    for (int j = 0; j < num_pred; j++) {
        double x = valores_pred[j];
        double y = b0 + b1 * x;
        printf("  Inversion = %5.0f  =>  Ventas estimadas = %.2f\n", x, y);
    }

    free(datos);
    return 0;
}

File: Regresion_Lineal/regresion_lineal.py
import sys
import io
import os
import math
import mysql.connector

sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8", errors="replace")
sys.stderr = io.TextIOWrapper(sys.stderr.buffer, encoding="utf-8", errors="replace")

def conectar():
    # Credenciales tomadas de variables de entorno (con valores por defecto
    # para el entorno local de XAMPP). Evita exponer contrasenas en el codigo.
    return mysql.connector.connect(
        host=os.getenv("DB_HOST", "127.0.0.1"),
        port=int(os.getenv("DB_PORT", "3306")),
        user=os.getenv("DB_USER", "root"),
        password=os.getenv("DB_PASS", ""),
        database=os.getenv("DB_NAME", "mineria_datos")
    )

def obtener_datos(cursor):
    cursor.execute("SELECT mes, inversion, ventas FROM inversion_ventas ORDER BY id")
    return cursor.fetchall()

def regresion_lineal(datos):
    n = len(datos)
    x = [fila[1] for fila in datos]
    y = [fila[2] for fila in datos]

    suma_x  = sum(x)
    suma_y  = sum(y)
    suma_xy = sum(xi * yi for xi, yi in zip(x, y))
    suma_x2 = sum(xi ** 2 for xi in x)

    b1 = (n * suma_xy - suma_x * suma_y) / (n * suma_x2 - suma_x ** 2)
    b0 = (suma_y - b1 * suma_x) / n

    return b0, b1

def metricas(datos, b0, b1):
    n = len(datos)
    y = [fila[2] for fila in datos]
    y_media = sum(y) / n

    ss_tot = sum((yi - y_media) ** 2 for yi in y)
    ss_res = sum((fila[2] - (b0 + b1 * fila[1])) ** 2 for fila in datos)

    r2   = 1 - ss_res / ss_tot if ss_tot > 0 else 1.0
    mse  = ss_res / n
    rmse = math.sqrt(mse)
    return r2, mse, rmse

def predecir(b0, b1, x):
    return b0 + b1 * x

def main():
    conn = conectar()
    cursor = conn.cursor()

    datos = obtener_datos(cursor)

    print("=" * 40)
    print("  DATOS DE LA BASE DE DATOS")
    print("=" * 40)
    print(f"{'Mes':<12} {'Inversion (X)':>14} {'Ventas (Y)':>12}")
    print("-" * 40)
    for mes, inv, ven in datos:
        print(f"{mes:<12} {float(inv):>14.2f} {float(ven):>12.2f}")

    b0, b1 = regresion_lineal(datos)

    print("\n" + "=" * 40)
    print("  REGRESION LINEAL")
    print("=" * 40)
    print(f"Intercepto  (b0): {b0:.4f}")
    print(f"Pendiente   (b1): {b1:.4f}")
    print(f"Ecuacion:  Y = {b0:.2f} + {b1:.2f} * X")

    r2, mse, rmse = metricas(datos, b0, b1)
    print("\n" + "=" * 40)
    print("  METRICAS DE BONDAD DE AJUSTE")
    print("=" * 40)
    print(f"R^2  (determinacion): {r2:.4f}  ({float(r2) * 100:.2f}% explicado)")
    print(f"MSE  (error cuad. medio): {mse:.4f}")
    print(f"RMSE (raiz del MSE): {rmse:.4f}")

    print("\n" + "=" * 40)
    print("  PREDICCIONES")
    print("=" * 40)
    valores_x = [row[1] for row in datos] + [40, 50, 60]
    for x_pred in valores_x:
        y_pred = predecir(b0, b1, x_pred)
        print(f"  Inversion = {float(x_pred):>5.0f}  =>  Ventas estimadas = {y_pred:.2f}")

    cursor.close()
    conn.close()

if __name__ == "__main__":
    main()
